client/main.js

```diff
@@ -0,0 +1,549 @@
+  this.sendTransaction = function sendTransaction(transaction) {
+    const tx = this.buildTx(transaction);
+    return signer.provider.send("eth_sendTransaction", [tx]);
+  };
+
+  this.estimateGas = async function estimateGas(transaction) {
+    const tx = this.buildTx(transaction);
+    const res = await signer.provider.send("eth_estimateGas", [tx]);
+    return ethers.utils.bigNumberify(res);
+  };
+
+  this.buildTx = function buildTx(transaction) {
+    const tx = {
+      from: this.address
+    };
+    ["to", "data"].forEach(key => {
+      if (transaction[key] != null) {
+        tx[key] = transaction[key];
+      }
+    });
+    ["gasPrice", "nonce", "value"].forEach(key => {
+      if (transaction[key] != null) {
+        tx[key] = ethers.utils.hexlify(transaction[key]);
+      }
+    });
+    if (transaction.gasLimit != null) {
+      tx.gas = ethers.utils.hexlify(transaction.gasLimit);
+    }
+    return tx;
+  };
+}
+// ===================================
+
+class ReputationMiningClient {
+  /**
+   * Constructor for ReputationMiningClient
+   * @param {string} minerAddress            The address that is staking CLNY that will allow the miner to submit reputation hashes
+   * @param {Number} [realProviderPort=8545] The port that the RPC node with the ability to sign transactions from `minerAddress` is responding on. The address is assumed to be `localhost`.
+   */
+  constructor(minerAddress, realProviderPort = 8545) {
+    this.minerAddress = minerAddress;
+    const ganacheProvider = ganache.provider({
+      network_id: 515,
+      vmErrorsOnRPCResponse: false,
+      locked: false,
+      verbose: true,
+      accounts: [
+        {
+          balance: "0x10000000000000000000000000",
+          secretKey
+        }
+      ]
+    });
+    this.ganacheProvider = new ethers.providers.Web3Provider(ganacheProvider);
+    this.ganacheWallet = new ethers.Wallet(secretKey, this.ganacheProvider);
+
+    this.realProvider = new ethers.providers.JsonRpcProvider(`http://localhost:${realProviderPort}`);
+    this.realWallet = new MetamaskSigner(minerAddress, this.realProvider);
+
+    try {
+      this.reputations = jsonfile.readFileSync(file);
+    } catch (err) {
+      this.reputations = {};
+    }
+  }
+
+  /**
+   * Initialises the mining client so that it knows where to find the `ColonyNetwork` contract
+   * @param  {string}  colonyNetworkAddress The address of the current `ColonyNetwork` contract
+   * @return {Promise}
+   */
+  async initialise(colonyNetworkAddress) {
+    const patriciaTreeDeployTx = ethers.Contract.getDeployTransaction(PatriciaTreeJSON.bytecode, PatriciaTreeJSON.abi);
+    const tx = await this.ganacheWallet.sendTransaction(patriciaTreeDeployTx);
+    this.reputationTree = new ethers.Contract(ethers.utils.getContractAddress(tx), PatriciaTreeJSON.abi, this.ganacheWallet);
+    this.nReputations = 0;
+    this.setColonyNetworkAddress(colonyNetworkAddress);
+  }
+
+  async setColonyNetworkAddress(address) {
```

Comment:
Do we need a dedicated function for this? Maybe push that one line up in `initialise`?